lib/srv/app/server.go

```diff
@@ -592,29 +601,112 @@ func (s *Server) HandleConnection(conn net.Conn) {
 }
 
 func (s *Server) handleConnection(conn net.Conn) error {
+	// Make sure everything here is wrapped in the tracking read connection for monitoring.
+	ctx, cancel := context.WithCancel(s.closeContext)
+	tc, err := srv.NewTrackingReadConn(srv.TrackingReadConnConfig{
+		Conn:    conn,
+		Clock:   s.c.Clock,
+		Context: ctx,
+		Cancel:  cancel,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
 	// Proxy sends a X.509 client certificate to pass identity information,
 	// extract it and run authorization checks on it.
-	tlsConn, user, app, err := s.getConnectionInfo(s.closeContext, conn)
+	tlsConn, user, app, err := s.getConnectionInfo(ctx, tc)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	authCtx, _, err := s.authorizeContext(context.WithValue(ctx, auth.ContextUser, user))
+	if err != nil {
+		// If the app is not TCP, write an HTTP error response and return nil instead of the error.
+		// This will ensure that HTTP clients will see a "403 Forbidden" code and message and not a 500.
+		if !app.IsTCP() {
+			code := trace.ErrorToCode(err)
+			resp := http.Response{
+				StatusCode: code,
+				Body:       io.NopCloser(bytes.NewReader([]byte(http.StatusText(code)))),
+			}
+			if writeErr := resp.Write(tlsConn); writeErr != nil {
+				s.log.Warnf("error writing HTTP response: %v", writeErr)
+			}
+
+			// Since we're not passing this connection to the http server, we should close
+			// the connection explicitly here.
+			if closeErr := conn.Close(); closeErr != nil {
+				s.log.Warnf("error closing connection: %v", closeErr)
+			}
+
+			return nil
```

Comment:
Why are we returning nil if there was an authz error?

Reply:
It's an issue of readability and response type. Here's the test code for AuthorizeWithLocks:

```
func TestAuthorizeWithLocks(t *testing.T) {
	s := SetUpSuite(t)
	// Create a lock targeting the user.
	lock, err := types.NewLock("test-lock", types.LockSpecV2{
		Target: types.LockTarget{User: s.user.GetName()},
	})
	require.NoError(t, err)
	s.tlsServer.Auth().UpsertLock(s.closeContext, lock)
	defer func() {
		s.tlsServer.Auth().DeleteLock(s.closeContext, lock.GetName())
	}()

	s.checkHTTPResponse(t, s.clientCertificate, func(resp *http.Response) {
		require.Equal(t, resp.StatusCode, http.StatusForbidden)
		buf, err := io.ReadAll(resp.Body)
		require.NoError(t, err)
		require.Equal(t, strings.TrimSpace(string(buf)), "Forbidden")
	})
}
```

So previously, what happens is that, for HTTP connections, no auth happens during this function, but auth happens during the actual `serveHTTP` function later. This allows app access to push out a `403 Forbidden`. If we return an error here, no valid HTTP status is sent to the HTTP client, so the client gets a `500` instead of the `403`. That test above actually fails without this behavior, and I wanted to maintain that behavior because of UX concerns.

Reply:
This also bypasses the tracing middleware. Maybe one alternative is to actually serve the original connection (e.g `s.handleHTTPApp(ctx, tlsConn)`) without monitor? we know it will fail anyway.

Reply:
Does this bypass the middleware? By tracing middleware, do you mean the `tc` connection?

Reply:
Sorry i wasn't being very clear. I meant httplib.MakeTracingHandler(
https://github.com/gravitational/teleport/blob/825c976968d126c0aef61c11079734338e8b841e/lib/srv/app/server.go#L844

so auth errors used to be traced if tracing is on. It's not a must-have but would be great if we can preserve that.

Reply:
I agree, I think we should preserve this too. I'll make sure we do.